src/Microsoft.DotNet.Build.Tasks/VersionTools/SubmitPullRequest.cs

```diff
@@ -78,19 +108,79 @@ private void TraceListenedExecute()
                     body += PullRequestCreator.NotificationString(NotifyGitHubUsers.Select(item => item.ItemSpec));
                 }
 
-                var prCreator = new PullRequestCreator(client.Auth, GitHubAuthor);
+                var options = new PullRequestOptions
+                {
+                    ForceCreate = AlwaysCreateNewPullRequest,
+                    MaintainersCanModify = MaintainersCanModifyPullRequest,
+                    TrackDiscardedCommits = TrackDiscardedCommits
+                };
+
+                var prCreator = new PullRequestCreator(client.Auth, PullRequestAuthor);
                 prCreator.CreateOrUpdateAsync(
                     CommitMessage,
                     CommitMessage + $" ({ProjectRepoBranch})",
                     body,
                     upstreamBranch,
                     origin,
-                    new PullRequestOptions
+                    options,
+                    client).Wait();
+            }
+        }
+
+        private IGitHubClient CreateClient(out GitHubProject origin)
+        {
+            PullRequestServiceType type;
+            if (!Enum.TryParse(PullRequestServiceType, true, out type))
+            {
+                string options = string.Join(", ", Enum.GetNames(typeof(PullRequestServiceType)));
+                throw new ArgumentException(
+                    $"{nameof(PullRequestServiceType)} '{PullRequestServiceType}' is not valid. " +
+                    $"Options are {options}");
+            }
+
+            switch (type)
+            {
+                case VersionTools.PullRequestServiceType.GitHub:
+                    origin = new GitHubProject(ProjectRepoName, PullRequestUser);
+                    return new GitHubClient(
+                        new GitHubAuth(
+                            PullRequestAuthToken,
+                            PullRequestUser,
+                            PullRequestEmail));
+
+                case VersionTools.PullRequestServiceType.Vsts:
+                    if (string.IsNullOrEmpty(VstsInstanceName))
+                    {
+                        throw new ArgumentException($"{nameof(VstsInstanceName)} is required but not set.");
+                    }
+
+                    // Get profile information from VSTS to use to create the real client.
+                    using (var idClient = new VstsAdapterClient(
+                        new GitHubAuth(PullRequestAuthToken),
+                        VstsInstanceName,
+                        VstsApiVersionOverride))
                     {
-                        ForceCreate = AlwaysCreateNewPullRequest,
-                        MaintainersCanModify = MaintainersCanModifyPullRequest,
-                        TrackDiscardedCommits = TrackDiscardedCommits
-                    }).Wait();
+                        VstsProfile profile = idClient.GetMyProfileAsync().Result;
+
+                        PullRequestAuthor = PullRequestUser;
```

Comment:
Why are we overwriting `PullRequestAuthor` here?

Reply:
Ah, yeah, this is because normally Author defaults to User based on the user in `GitHubAuth`, but for VSTS we set the `GitHubAuth` user to the user's VSTS GUID because that's used later for the PR search.

Looking at it again, I don't like how this odd bit of logic is in the build task. I think I can move it into the `PullRequestCreator` to make VSTS more consistent to use from the .NET lib.